### ISSUE TITLE
nodegroup deletion is blocked by SG/ENIs

### ISSUE THREAD

**Issue by martina-if** [weaveworks/eksctl#1849]
We have a failure in integration tests:

```
(Integration) Create, Get, Scale & Delete when creating a cluster with 1 node 
  should not return an error
  /src/integration/creategetdelete_test.go:62
Using kubeconfig: /src/integration/kubeconfig-451144843
starting '../eksctl "--region" "us-west-2" "create" "cluster" "--verbose" "4" "--name" "scrumptious-gopher-1582686390" "--tags" "alpha.eksctl.io/description=eksctl integration test" "--nodegroup-name" "ng-0" "--node-labels" "ng-name=ng-0" "--nodes" "1" "--version" "1.14" "--kubeconfig" "/src/integration/kubeconfig-451144843"'
.
.
.
2020-02-26T05:15:04Z [✖]  unexpected status "DELETE_FAILED" while waiting for CloudFormation stack "eksctl-managed-hilarious-wardrobe-1582686390-nodegroup-managed-ng-0"
2020-02-26T05:15:04Z [ℹ]  fetching stack events in attempt to troubleshoot the root cause of the failure
2020-02-26T05:15:04Z [✖]  AWS::CloudFormation::Stack/eksctl-managed-hilarious-wardrobe-1582686390-nodegroup-managed-ng-0: DELETE_FAILED – "The following resource(s) failed to delete: [SG]. "
2020-02-26T05:15:04Z [✖]  AWS::EC2::SecurityGroup/SG: DELETE_FAILED – "resource sg-0161c1e34bd99c2e6 has a dependent object (Service: AmazonEC2; Status Code: 400; Error Code: DependencyViolation; Request ID: 80f4c7fb-1b9a-4fdc-b6ab-fd4e7ef5fdf8)"
```
https://app.circleci.com/jobs/github/weaveworks/eksctl/6620/parallel-runs/0/steps/0-107

**Comment by lliknart** [comment]
Hi, 

I face to the same error using eksctl and managed node groups

```
eksctl delete cluster -f my-cluster.yml

[ℹ]  eksctl version 0.14.0
[ℹ]  using region eu-west-3
[ℹ]  deleting EKS cluster "my-cluster-dev"
[ℹ]  either account is not authorized to use Fargate or region eu-west-3 is not supported. Ignoring error
[✔]  kubeconfig has been updated
[ℹ]  cleaning up LoadBalancer services
[ℹ]  2 sequential tasks: { 2 parallel sub-tasks: { cleanup for nodegroup "managed-ng-1", delete nodegroup "managed-ng-1" }, delete cluster control plane "my-cluster-dev" [async] }
[ℹ]  trying to cleanup dangling network interfaces
[ℹ]  will delete stack "my-cluster-dev-nodegroup-managed-ng-1"
[ℹ]  waiting for stack "my-cluster-dev-nodegroup-managed-ng-1" to get deleted
[✖]  unexpected status "DELETE_FAILED" while waiting for CloudFormation stack "my-cluster-dev-nodegroup-managed-ng-1"
[ℹ]  fetching stack events in attempt to troubleshoot the root cause of the failure
[✖]  AWS::CloudFormation::Stack/my-cluster-dev-nodegroup-managed-ng-1: DELETE_FAILED – "The following resource(s) failed to delete: [ManagedNodeGroup]. "
[✖]  AWS::EKS::Nodegroup/ManagedNodeGroup: DELETE_FAILED – "Nodegroup managed-ng-1 failed to stabilize: Internal Failure"
[✖]  AWS::CloudFormation::Stack/my-cluster-dev-nodegroup-managed-ng-1: DELETE_FAILED – "The following resource(s) failed to delete: [ManagedNodeGroup]. "
[✖]  AWS::EKS::Nodegroup/ManagedNodeGroup: DELETE_FAILED – "Nodegroup managed-ng-1 failed to stabilize: Internal Failure"
[ℹ]  1 error(s) occurred while deleting cluster with nodegroup(s)
[✖]  waiting for CloudFormation stack "my-cluster-dev-nodegroup-managed-ng-1": ResourceNotReady: failed waiting for successful resource state
Error: failed to delete cluster with nodegroup(s)
```

I get this error on AWS console:

![image](https://user-images.githubusercontent.com/11667671/75981926-d0b0f080-5ee5-11ea-843f-cd1212e2152b.png)

**Comment by sayboras** [comment]
I am facing the same issue issue with un-managed node group as well.

<details>
<summary>Configuration</summary>

```yaml
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig
metadata:
  name: test-delete
  region: ap-southeast-2

nodeGroups:
  - name: standard-workers
    instanceType: t3.medium
    desiredCapacity: 1
    minSize: 1
    maxSize: 2
  - name: private-workers
    instanceType: t3.medium
    privateNetworking: true
    desiredCapacity: 1
    minSize: 1
    maxSize: 2
managedNodeGroups:
  - name: managed-workers
    instanceType: t3.medium
    desiredCapacity: 1
    minSize: 1
    maxSize: 2

```


</details>

<details>
<summary>Delete output</summary>

```
./eksctl delete cluster -f test_private.yaml --wait
[ℹ]  eksctl version 0.16.0-dev+239507a8.2020-03-09T20:24:27Z
[ℹ]  using region ap-southeast-2
[ℹ]  deleting EKS cluster "test-delete"
[ℹ]  either account is not authorized to use Fargate or region ap-southeast-2 is not supported. Ignoring error
[✔]  kubeconfig has been updated
[ℹ]  cleaning up LoadBalancer services
[ℹ]  2 sequential tasks: { 6 parallel sub-tasks: { cleanup for nodegroup "standard-workers", delete nodegroup "standard-workers", cleanup for nodegroup "private-workers", delete nodegroup "private-workers", cleanup for nodegroup "managed-workers", delete nodegroup "managed-workers" }, delete cluster control plane "test-delete" }
[ℹ]  trying to cleanup dangling network interfaces
[ℹ]  trying to cleanup dangling network interfaces
[ℹ]  trying to cleanup dangling network interfaces
[ℹ]  will delete stack "eksctl-test-delete-nodegroup-managed-workers"
[ℹ]  waiting for stack "eksctl-test-delete-nodegroup-managed-workers" to get deleted
[ℹ]  will delete stack "eksctl-test-delete-nodegroup-standard-workers"
[ℹ]  waiting for stack "eksctl-test-delete-nodegroup-standard-workers" to get deleted
[ℹ]  will delete stack "eksctl-test-delete-nodegroup-private-workers"
[ℹ]  waiting for stack "eksctl-test-delete-nodegroup-private-workers" to get deleted
[✖]  unexpected status "DELETE_FAILED" while waiting for CloudFormation stack "eksctl-test-delete-nodegroup-standard-workers"
[ℹ]  fetching stack events in attempt to troubleshoot the root cause of the failure
[✖]  AWS::CloudFormation::Stack/eksctl-test-delete-nodegroup-standard-workers: DELETE_FAILED – "The following resource(s) failed to delete: [SG]. "
[✖]  AWS::EC2::SecurityGroup/SG: DELETE_FAILED – "resource sg-0badfc766ec4a0369 has a dependent object (Service: AmazonEC2; Status Code: 400; Error Code: DependencyViolation; Request ID: 88abb1ef-d5f5-4cfb-aefb-c5e3bd3f9b79)"
[ℹ]  1 error(s) occurred while deleting cluster with nodegroup(s)
[✖]  waiting for CloudFormation stack "eksctl-test-delete-nodegroup-standard-workers": ResourceNotReady: failed waiting for successful resource state
```

</details>

**Comment by sayboras** [comment]
Maybe that is why we have condition for failed to delete stack https://github.com/weaveworks/eksctl/blob/master/pkg/cfn/manager/delete_tasks.go#L79, however will it better to avoid such delete failure in the first place (unless we can't) ?

**Comment by danmoore2205** [comment]
I had encountered the same issue...

```
$ eksctl delete cluster --config-file aws_cluster.yaml
[ℹ]  eksctl version 0.16.0
[ℹ]  using region ap-southeast-2
[ℹ]  deleting EKS cluster "learn-kube"
[ℹ]  either account is not authorized to use Fargate or region ap-southeast-2 is not supported. Ignoring error
[✔]  kubeconfig has been updated
[ℹ]  cleaning up LoadBalancer services
[ℹ]  2 sequential tasks: { 2 parallel sub-tasks: { cleanup for nodegroup "ng-1", delete nodegroup "ng-1" }, delete cluster control plane "learn-kube" [async] }
[ℹ]  trying to cleanup dangling network interfaces
[ℹ]  will delete stack "eksctl-learn-kube-nodegroup-ng-1"
[ℹ]  waiting for stack "eksctl-learn-kube-nodegroup-ng-1" to get deleted
[✖]  unexpected status "DELETE_FAILED" while waiting for CloudFormation stack "eksctl-learn-kube-nodegroup-ng-1"
[ℹ]  fetching stack events in attempt to troubleshoot the root cause of the failure
[✖]  AWS::CloudFormation::Stack/eksctl-learn-kube-nodegroup-ng-1: DELETE_FAILED – "The following resource(s) failed to delete: [ManagedNodeGroup]. "
[✖]  AWS::EKS::Nodegroup/ManagedNodeGroup: DELETE_FAILED – "Nodegroup ng-1 failed to stabilize: Internal Failure"
[✖]  AWS::CloudFormation::Stack/eksctl-learn-kube-nodegroup-ng-1: DELETE_FAILED – "The following resource(s) failed to delete: [ManagedNodeGroup]. "
[✖]  AWS::EKS::Nodegroup/ManagedNodeGroup: DELETE_FAILED – "Nodegroup ng-1 failed to stabilize: Internal Failure"
[✖]  AWS::CloudFormation::Stack/eksctl-learn-kube-nodegroup-ng-1: DELETE_FAILED – "The following resource(s) failed to delete: [ManagedNodeGroup]. "
[✖]  AWS::EKS::Nodegroup/ManagedNodeGroup: DELETE_FAILED – "Nodegroup ng-1 failed to stabilize: Internal Failure"
[✖]  AWS::CloudFormation::Stack/eksctl-learn-kube-nodegroup-ng-1: DELETE_FAILED – "The following resource(s) failed to delete: [NodeInstanceRole, ManagedNodeGroup]. "
[✖]  AWS::IAM::Role/NodeInstanceRole: DELETE_FAILED – "Cannot delete entity, must detach all policies first. (Service: AmazonIdentityManagement; Status Code: 409; Error Code: DeleteConflict; Request ID: 35067978-5029-4e0b-a4be-c74030e1590b)"
[✖]  AWS::EKS::Nodegroup/ManagedNodeGroup: DELETE_FAILED – "Nodegroup ng-1 failed to stabilize: Internal Failure"
[ℹ]  1 error(s) occurred while deleting cluster with nodegroup(s)
[✖]  waiting for CloudFormation stack "eksctl-learn-kube-nodegroup-ng-1": ResourceNotReady: failed waiting for successful resource state
Error: failed to delete cluster with nodegroup(s)
```

Followed step 2 above from @lliknart and then `eksctl delete cluster` worked successfully. I too had a dangling ENI(that was showing as available).

**Comment by pierluigilenoci** [comment]
I've exactly the same issue. I also opened an internal support ticket. Let's see the outcome.

**Comment by JoshuaCWebDeveloper** [comment]
Had the exact same problem. Tried re-running the delete, did not work. Deleting the dangling ENI *did* work. Here is how you can delete the ENI using the console:
1. Go to **EC2 > Network Interfaces**
2. Sort by VPC, find the interfaces assigned to your VPC
3. The interface to delete should be the only one that is "available", it should also be the only one assigned to the problematic remote access SG.

**Comment by 1beb** [comment]
Still happening. Exact same issue. Fails on delete, leaving a bunch of resources left over. Complains that it can't delete:

```
AWS::CloudFormation::Stack/eksctl-matrix-nodegroup-standard-workers: DELETE_FAILED – "The following resource(s) failed to delete: [SSH]. "
2021-07-18 01:28:10 [✖]  AWS::EC2::SecurityGroup/SSH: DELETE_FAILED – "resource sg-XXXX has a dependent object (Service: AmazonEC2; Status Code: 400; Error Code: DependencyViolation; Request ID: XXXX; Proxy: null)"
```

And there are dangling ENI devices that don't want to be deleted because they are "in use"

**Comment by nikimanoledaki** [comment]
Thanks for the update @1beb. Is this something you still come across? Does deleting again with `--wait` solve the issue for you?

**Comment by khaled** [comment]
@nikimanoledaki I'm not sure if this is related or not because I am not using managed node groups, but I've seen a few delete failures recently where SG's would not delete.  Deleting again with `---wait` does not help.  Here's an example log:

```
AWS::CloudFormation::Stack/eksctl-XYZ-nodegroup-ng-1: DELETE_FAILED – "The following resource(s) failed to delete: [SG]. "
2021-09-07 11:04:41 [✖]  AWS::EC2::SecurityGroup/SG: DELETE_FAILED – "resource sg-XYZ has a dependent object (Service: AmazonEC2; Status Code: 400; Error Code: DependencyViolation; Request ID: XYZ; Proxy: null)"
```

The dependent object in question appears to be a network interface.

**Comment by nikimanoledaki** [comment]
Thanks @khaled, I'm trying to reproduce this so any info helps! 

Could you please share a bit more info about your configuration (your config file if you used one, with sensitive info removed)?  

If you know the steps for how to reproduce this error and could share these, that would be ideal. Otherwise I'll investigate. Thanks!

**Comment by khaled** [comment]
@nikimanoledaki I don't know if this will be terribly interesting, but here's our config with some bits redacted:

```yaml
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig
metadata:
  name: my-cluster
  region: <redacted>
  tags:
    alpha.eksctl.io/cluster-name: my-cluster
vpc: <redacted; uses an existing vpc>
nodeGroups:
  - name: my-ng
    instanceType: m5.large
    desiredCapacity: 2
    minSize: 0
    maxSize: 3
    volumeSize: 40
    volumeEncrypted: true
    availabilityZones: [ <redacted> ]
    privateNetworking: true
    labels:
      <redacted>
    iam:
      attachPolicyARNs:
        - arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy
        - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
        - <redacted>
      withAddonPolicies:
        imageBuilder: true
        cloudWatch: true
        autoScaler: true
    preBootstrapCommands:
      - <redacted; logic here to tag root volumes>
 ```

**Comment by nikimanoledaki** [comment]
Closing this since it should be solved by #4214, which will drain the nodegroups and delete Daemonsets (e.g. VPC CNI!) to release networking-related resources (SGs, ENIs, IPs) before deleting the cluster. This is part of the braoder process of upgrading our cluster/nodegroup deletion strategy (see #536).
